### DISCUSSION TITLE
How do update a variable periodically ?

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#10187] [Q&A]
[_From Discord_:](https://discord.com/channels/372137812037730304/550416151172087808/1032003988142432276)

I have spent 2 days trying to update a variable cyclically in the UI through binding and I've had no success. I am using the UNO.Extensions template that targets all platforms and includes a 2nd page.

1. View - ViewModel Mapping:
```
.UseNavigation(ReactiveViewModelMappings.ViewModelMappings, RegisterRoutes)
```
Where
```
views.Register(new ViewMap<MainPage, MainViewModel>());
```
3. ViewModel Inheritance
```
public partial class MainViewModel : ObservableObject
```

5. ViewModel Observable Property
```
[ObservableProperty]
private int _sampleRate;
```

7. Update Task
```
public async Task UpdateSampleRateAsync()
    {
        try
        {
            while (true)
            {
                SampleRate = _mcService?.GetSampleRate() ?? -1;
                await Task.Delay(1);
            }
        }
        catch (Exception ex)
        {
            var e = ex;
        }
    }
```

9. Binding
```
<TextBlock Text="{Binding SampleRate}" />
```

or 
```
<TextBlock Text="{x:Bind ViewModel.SampleRate}" />
```
where:
```
public sealed partial class MainPage : Page
{
    public MainViewModel? ViewModel => DataContext as MainViewModel;
```

nothing has worked. and my hot reload also does not work. all i did was create a template with uno command and using latest VS22

**Comment by jpgarza93** [comment]
I created an "Uno Platform App" project from Visual Studio 22 (v17.3.6) and everything worked as expected:

1. x:Bind
2. Hot Reload

I am just not using the UNO Extensions template:
```
.UseNavigation(ReactiveViewModelMappings.ViewModelMappings, RegisterRoutes)
views.Register(new ViewMap<MainPage, MainViewModel>());
```

**Reply by dr1rrb** [reply]
Your VM is using MVVM toolkit, so you don't need to use the `ReactiveViewModelMappings.ViewModelMappings`. 

When you use it, instead of data-binding to your VM itself, you are data-bound to the generated "Bindable" version of your VM (which re-exposes the properties of your VM), but so far we are not forwarding the `INotifyPropertyChange` events from your VM to the generated BindableVM (cf. [this issue](https://github.com/unoplatform/uno.extensions/issues/617)).

The `BindableVM` is for usage with `Feed`, if you want you can do:
```csharp
public Feed<int> SampleRate => Feed.AsyncEnumerable(GetSampleRateAsync);

private async IAsyncEnumerable<int> GetSampleRateAsync([EnumeratorCancellation] CancellationToken ct = default)
{
    try
    {
        while (!ct.IsCancellationRequested)
        {
            yield return _mcService?.GetSampleRate() ?? -1;
            await Task.Delay(1);
        }
    }
    catch (Exception ex)
    {
        var e = ex;
    }
}
```

**Reply by francoistanguay** [reply]
@jpgarza93  Do you have a sample repro app you can share?

**Reply by jpgarza93** [reply]
@dr1rrb thank you. It is hard to understand what is going on View <-> ViewModel wise in the UNO.Extensions template. Not many comments on what is happening IMO. 

@francoistanguay I do not currently. but I can share the small project I was working on if needed. If you need it, would you like me to just share a zip file? or how should I share it?

**Reply by francoistanguay** [reply]
Can you please try to remove the 1st parameter in UseNavigation? You're using MVVM Toolkit and therefore dont need to rely on Extensions.Reactive

.UseNavigation(ReactiveViewModelMappings.ViewModelMappings, RegisterRoutes)

to

.UseNavigation(RegisterRoutes)

**Reply by francoistanguay** [reply]
And yes you can add an attachment if my suggestion didnt fix your issue

**Reply by jpgarza93** [reply]
Removing this `ReactiveViewModelMappings.ViewModelMappings` did not fix the issue

**Reply by jpgarza93** [reply]
My bindable property updates correctly one time, but then the next time it tries to assign the value again I get this error:

`{"The application called an interface that was marshalled for a different thread. (0x8001010E (RPC_E_WRONG_THREAD))"}`

**Reply by dr1rrb** [reply]
Well this error means that you are trying to update the view from a background thread. Looking at your code, the most probable reason is that you are updating the `SampleRate` from a background thread, which will raise the `NotifyPropertyChanged` from that background thread (I think that the MVVM toolkit does not handle multi-threading : it's your responsibility to make sure to go back on the UI thread before updating the `SampleRate`).

Note: This could be achieve automatically by the system if you make sure to invoke your `public async Task UpdateSampleRateAsync()` from the UI thread.

You can request for an `IDispatcher` in your VM constructor do to that.
